packages/rest/src/router/routing-table.ts

```diff
@@ -259,55 +342,75 @@ export class Route extends BaseRoute {
   }
 }
 
-type ControllerInstance = {[opName: string]: Function};
-
-export class ControllerRoute extends BaseRoute {
+/**
+ * A route backed by a controller
+ */
+export class ControllerRoute<T> extends BaseRoute {
+  protected readonly _controllerCtor: ControllerClass<T>;
+  protected readonly _controllerName: string;
   protected readonly _methodName: string;
-
+  protected readonly _controllerFactory: ControllerFactory<T>;
+
+  /**
+   * Construct a controller based route
+   * @param verb http verb
+   * @param path http request path
+   * @param spec OpenAPI operation spec
+   * @param controllerCtor Controller class
+   * @param controllerFactory A factory function to create a controller instance
+   * @param methodName Controller method name, default to `x-operation-name`
+   */
   constructor(
     verb: string,
     path: string,
     spec: OperationObject,
-    protected readonly _controllerCtor: ControllerClass,
+    controllerCtor: ControllerClass<T>,
+    controllerFactory?: ControllerFactory<T>,
     methodName?: string,
   ) {
+    const controllerName = spec['x-controller-name'] || controllerCtor.name;
+    methodName = methodName || spec['x-operation-name'];
+
+    if (!methodName) {
+      throw new Error(
+        'methodName must be provided either via the ControllerRoute argument ' +
+          'or via "x-operation-name" extension field in OpenAPI spec. ' +
+          `Operation: "${verb} ${path}" ` +
+          `Controller: ${controllerName}.`,
+      );
+    }
+
     super(
       verb,
       path,
       // Add x-controller-name and x-operation-name if not present
       Object.assign(
         {
-          'x-controller-name': _controllerCtor.name,
+          'x-controller-name': controllerName,
           'x-operation-name': methodName,
-          tags: [_controllerCtor.name],
+          tags: [controllerName],
         },
         spec,
       ),
     );
 
-    if (!methodName) {
-      methodName = this.spec['x-operation-name'];
-    }
-
-    if (!methodName) {
-      throw new Error(
-        'methodName must be provided either via the ControllerRoute argument ' +
-          'or via "x-operation-name" extension field in OpenAPI spec. ' +
-          `Operation: "${verb} ${path}" ` +
-          `Controller: ${this._controllerCtor.name}.`,
-      );
-    }
-
+    this._controllerFactory =
+      controllerFactory || createControllerFactoryForClass(controllerCtor);
```

Comment:
This is the part I'd prefer to avoid - can we modify `ControllerRoute` to require the factory argument, and let `app.controller` and friends to supply the right value?